src/docs/binding-generics.md

```diff
@@ -437,3 +437,190 @@ We could also consider solving this on the runtime level.
+        PInvokePrintMe(self);
+    }
+}
+```
+
+### Protocol constraints on types
+
+Adding constraints on a type's generic parameters is also straightforward, with small adjustments necessary to the initializer and metadata accessor functions. Let's revisit the `Pair` example:
+
+```swift
+public struct Pair<T: Printable, U: Printable> { 
+    public var first: T 
+    public var second: U
+ 
+    public init(first: T, second: U) { 
+        self.first = first 
+        self.second = second 
+    }
+
+    public func printMe() {
+        self.first.printMe()
+        self.second.printMe()
+    }
+}
+```
+
+The corresponding C\# projection would be:
+
+```csharp
+class Pair<T, U> : ISwiftObject
+    where T : IPrintableProtocol
+    where U : IPrintableProtocol
+{
+
+    private static nuint PayloadSize =  /* Extract size */;
+    private NativeHandle _payload;
+
+    public unsafe Pair(T first, U second)
+    {
+        _payload = new NativeHandle(NativeMemory.Alloc(PayloadSize));
+
+        SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(_payload);
+
+        var firstMetadata = GetTypeMetadataOrThrow<T>();
+        var secondMetadata = GetTypeMetadataOrThrow<U>();
+
+        var nativeHandleFirst = Runtime.GetPayload(ref first);
+        var nativeHandleSecond = Runtime.GetPayload(ref second);
+
+        var printableProtocolWitnessTableT = T.GetIPrintableProtocolWitnessTable();
+        var printableProtocolWitnessTableU = U.GetIPrintableProtocolWitnessTable();
+
+        PairPInvokes.Pair(swiftIndirectResult, nativeHandleFirst, nativeHandleSecond, firstMetadata, secondMetadata, printableProtocolWitnessTableT, printableProtocolWitnessTableU);
+    }
+
+    public static TypeMetadata Metadata
+    {
+        /* This should be cached */
+        get
+        {
+            var firstMetadata = GetTypeMetadataOrThrow<T>();
+            var secondMetadata = GetTypeMetadataOrThrow<U>();
+
+            var printableProtocolWitnessTableT = T.GetIPrintableProtocolWitnessTable();
+            var printableProtocolWitnessTableU = U.GetIPrintableProtocolWitnessTable();
+
+            // Might be handled differently due to the metadata lowering https://github.com/dotnet/runtimelab/pull/2810
+
+            return PairPInvokes.PInvokeMetadata(TypeMetadataRequest.Complete, firstMetadata, secondMetadata, printableProtocolWitnessTableT, printableProtocolWitnessTableU);
+        }
+    }
+
+    // Other members
+}
+```
+
+### Supporting generic constraints on types present in both C\# and Swift
+
+When projecting a Swift type into C\#, we can project all of the protocols it conforms to as interfaces and make the C\# equivalent implement those interfaces. This allows it to work with the generic constraints mechanism described above.
+
+However, types that exist in both languages require special handling—for example, primitive number types. In Swift, primitive number types conform to multiple protocols that their C\# counterparts do not.
```

Comment:
Do I understand it correctly that plain Swift primitive types, e.g., Swift [Int](https://developer.apple.com/documentation/swift/int) conform, by default, to multiple protocols?

Reply:
This is absolutely correct. Two things to get used to with Swift as a language is (1) they've done their best to hide that there even are primitive types. For the most part, all the integral types, floating point types, etc are just value types and you can do anything to them that you can do to any other value type. (2) extensions allow any type to implement nearly* any other protocol.

*nearly - in that some PATs can have constraints on them to make them incompatible with value types